### PR TITLE
drnim: tiny progress

### PR DESCRIPTION
**Pull request by Araq**

### CODE REVIEW

**Comment by juancarlospaco** [comment]
@Araq 
Should all pragmas be Plural?, `invariant` :arrow_right: `invariants`,
because `invariant` makes you think it only takes 1 per pragma which is wrong,  not too late to fix.
:thinking:

**Comment by Araq** [comment]
But it only takes one invariant per pragma. I don't understand your question.

**Comment by juancarlospaco** [comment]
Ok, I am wrong, I get confused the example is using `forall()`, makes sense.

**Comment by juancarlospaco** [comment]
Can it be like?, maybe overload `forall()` to take arguments like this?:
```nim
{.requires: forall(
  (argument > 0, "argument is wrong bro"), # Print this message too if condition fails
  (argument < 9, "argument is too big yo"), # Human-readable message to print on error
).}   

{.ensures: forall(
  (result > 0, "Oh no, you just crashed prod"), 
  (result < 9, "result must be < 9"), # Tuple[conditional: bool, message: string]
).}
```

**Comment by Araq** [comment]
This is not ``forall`` and the annotations are not "design by contract".

File: compiler/guards.nim [hunk 1]
```diff
@@ -482,8 +482,11 @@ proc invalidateFacts*(m: var TModel, n: PNode) =
   # The same mechanism could be used for more complex data stored on the heap;
   # procs that 'write: []' cannot invalidate 'n.kind' for instance. In fact, we
   # could CSE these expressions then and help C's optimizer.
-  for i in 0..high(m.s):
-    if m.s[i] != nil and m.s[i].hasSubTree(n): m.s[i] = nil
+  for i in 0..high(s):
+    if s[i] != nil and s[i].hasSubTree(n): s[i] = nil
```
Comment (timotheecour) [inline compiler/guards.nim:486]:
```
  for a in mitems(s):
    if a != nil and a.hasSubTree(n): a = nil
```

File: drnim/drnim.nim [hunk 2]
```diff
@@ -564,12 +568,299 @@ proc compatibleProps(graph: ModuleGraph; formal, actual: PType): bool {.nimcall.
       # we already know from the type system?
       result = frequires.isEmpty and fensures.isEmpty
 
+type
+  DrnimContext = ref object
+    graph: ModuleGraph
+    facts: seq[PNode]
+    o: Operators
+    hasUnstructedCf: int
```
Comment (timotheecour) [inline drnim/drnim.nim]: written but never read, and its name isn't clear, at least deserves a doc comment

File: drnim/drnim.nim [hunk 3]
```diff
@@ -564,12 +568,299 @@ proc compatibleProps(graph: ModuleGraph; formal, actual: PType): bool {.nimcall.
...
+    if n.len == 1:
+      traverse(c, n[0])
+      if optStaticBoundsCheck in c.currOptions:
+        checkRange(c, n[0], n.typ)
+  of nkBracketExpr:
+    if optStaticBoundsCheck in c.currOptions and n.len == 2:
+      if n[0].typ != nil and skipTypes(n[0].typ, abstractVar).kind != tyTuple:
+        checkBounds(c, n[0], n[1])
+    for i in 0 ..< n.len: traverse(c, n[i])
+  else:
```
Comment (timotheecour) [inline drnim/drnim.nim:1020]: `else:` +  `safeLen` is error prone, you'll never know whether you covered all the interesting cases; instead how about:
```nim
elif n in nkContainer: ...
else: doAssert false, $kind
```
or if you want to ignore specific ones:
```nim
else: doAssert kind in {whitelist...}, $kind
```
where `nkContainer = TNodeKind - {nkSym,nkIdent,nkNilLit} - nkLiterals - {}` is generally useful for exactly that kind of pattern

File: compiler/dfa.nim [hunk 4]
```diff
@@ -56,6 +56,7 @@ type
     tryStmtFixups: seq[TPosition]
     forks: seq[TPosition]
     owner: PSym
+    bePrecise*: bool
```
Comment (timotheecour) [inline compiler/dfa.nim]: unused

File: drnim/drnim.nim (hunk 2 above)
Comment (Araq) [reply drnim/drnim.nim]: It'll be read. This PR is WIP.

File: drnim/drnim.nim (hunk 3 above)
Comment (Araq) [reply drnim/drnim.nim:1020]: Ok.

**Comment by juancarlospaco** [comment]
Ok, I understand that this is not "design by contract"-like annotations,
but when finished, it may allow such feature?,
it is kinda like +90% whats needed to have it; But not on the compiler!,
on the tool or by third party module somehow (?)
is just a question :thinking:

**Comment by Araq** [comment]
> but when finished, it may allow such feature?,

Drnim is about proving your code correct, not about more fancy ways to write ``assert``.

**Comment by Araq** [comment]
Unrelated CI failure, merging.

**Pull request merged**